documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_moderation_metrics.md

```diff
@@ -0,0 +1,327 @@
+  end
+  subgraph a[Realtime metrics]
+    2[Step 2: Implement real-time metrics] --- 3[Step 3: Surface in AWS CloudWatch]
+  end
+  1[Step 1: Identify and categorise metrics] --- 2 & 4
+```
+
+In these steps, step 1 should be done sufficiently by this plan itself.
+
+Steps 2 and 4 need step 1 to be completed but can be done independently of each
+other. Steps 3 and 5 depends on 2 and 4 respectively. Hence, there can be two
+parallel tracks of work, one for the real time metrics and one for the deferred
+metrics.
+
+## Step details
+
+<!--
+Describe all of the implementation steps listed in the "step-by-step plan" in detail.
+
+For each step description, ensure the heading includes an obvious reference to the step as described in the
+"step-by-step plan" section above.
+-->
+
+```{caution}
+This document shows code-samples but intended as a guide and not meant to be
+used directly!
+```
+
+### Identify and categorise metrics
+
+The metrics we aim to measure can be classified along two axes:
+
+- Whether they measure user-submitted reports or moderator-generated responses
+- Whether they are real-time or deferred
+
+The following metrics can be considered to begin with (we can always add more in
+the future):
+
+- Metrics about reports
+  - number of reports (time-series, real-time)
+    <!-- TODO: - per media type (time-series, real-time) -->
+    - per violation (copyright, sensitive, other)
+  - most reported creator/source/provider (deferred)
+  - accuracy of reports (deferred)
+  - duplication in reports (deferred)
+- Metrics about responses
+  - number of decisions (time-series, real-time)
+    <!-- TODO: - per media type (time-series, real-time) -->
+  - number of resolutions (time-series, real-time)
+    <!-- TODO: - per media type (time-series, real-time) -->
+    - per resolution type (confirmation, rejection, deduplication)
+  - time to resolution (deferred)
+    - average
+    - P99
+
+The real-time metrics can go to a visual dashboard where they can be plotted as
+a time-series. They can be used to observe trends and spikes in reports (which
+is useful to understand usage and detect abuse of our reporting mechanisms) and
+in resolutions (which is useful to know how our resolution volume compares to
+reporting volume and when we need to scale our moderation resources).
+
+The deferred metrics are computed periodically from our database. They can be
+used to build reports that can be presented inside Django's admin interface
+(which are useful to identify the sources, providers and creators that have a
+lot of sensitive content and need bulk moderation). Having them in Django Admin
+brings them closer to where the moderation is being performed.
+
+### Implement real-time metrics
+
+Real-time metrics are implemented by emitting structured log lines from the
+moderation pipeline, whenever one of the following event occurs.
+
+- report created
+- report confirmed
+- report rejected
+- report deduplicated
+- decision created
```

Comment:
I think these events are a little confusing given the changes that have happened recently to the Report model. In the admin tools and access IP we decided to [deprecate the status column on MediaReports](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.html#deprecating-and-removing-report-status); it now has a `decision_id` linking an associated ModerationDecision. A Report is only ever 'pending' (if it does not have a `decision_id`) or 'reviewed' (if it does -- which may be rejected/deduplicated/marked_sensitive, etc).

A single decision can also be tied to multiple reports in the moderation workflow (see [IP](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.html#implementation-plan-django-admin-tools-and-access-control-for-moderators)). A decision also need not necessarily be linked to a report; for example, the decision to undo an action 🙃 

It's subtle but I think it would more accurately reflect the pipeline to have events like:

- report created
  - include the 'reason' 
- decision created
  - include the 'action_type'
- report reviewed
  - include the 'action_type'
  - this event is always a result of a decision being created. The `action_type` is the action taken by the decision.

So when a ModerationDecision is made, we would log events for:
- the decision itself being created
- each of the affected reports being reviewed

Reply:
That's a good catch, thank you! With e7eb22d, the events have been updated to reflect the updated usage of the report and decision models.

Events will now be logged when a report is created and when a decision is created. If the created decision is then associated with any number of reports, each of these reports will also a log an event. The schema of the events has also been updated.